File: Grammars/Table/TableGrammar.vb
Imports System.Reflection
Imports System.Text.RegularExpressions
Imports System.Xml.Serialization
Imports NCalc
Imports Utility
Imports Grammars.Table.TableGrammarTable
Imports System.Collections.ObjectModel

Namespace Table
    <XmlRoot([Namespace]:="", IsNullable:=False)>
    Public Class TableGrammar
        Inherits BaseGrammar

#Region "Constants"
        Const CALCULATION_REGEX As String = "\[=.*?\]"
        Const VARIABLE_OPEN As Char = "["c
        Const VARIABLE_CLOSE As Char = "]"c
        Const CALCULATION_OPEN As String = "[="
#End Region

#Region "Members"
        Private _valueList As New Dictionary(Of String, Object)
        Private WithEvents _calculator As New Calculator(Me)
#End Region

#Region "Public Properties"
        <XmlArray("tables")>
        <XmlArrayItem("table", IsNullable:=False)>
        Public Property Tables() As New ObservableCollection(Of TableGrammarTable)

        <XmlElement("output")>
        Public Property Output() As String
#End Region

#Region "Public Methods"
        Public Overrides Function Analyze() As String
            Return String.Empty
        End Function
#End Region

#Region "Protected Methods"
        Protected Overrides Function GenerateName() As String
            _valueList = New Dictionary(Of String, Object)
            For Each table As TableGrammarTable In Tables
                AddHandler table.EvaluateParameter, AddressOf Tables_EvaluateParameter
                AddHandler table.EvaluateFunction, AddressOf Tables_EvaluteFunction
                Dim newResults As Dictionary(Of String, Object) = table.ProcessTable()
                For Each result As KeyValuePair(Of String, Object) In newResults
                    _valueList.Add(table.Name & "." & result.Key, result.Value)
                Next
                RemoveHandler table.EvaluateParameter, AddressOf Tables_EvaluateParameter
                RemoveHandler table.EvaluateFunction, AddressOf Tables_EvaluteFunction
            Next
            Return FillOutput()
        End Function
#End Region

#Region "Private Methods"
        Private Function FillOutput() As String
            Dim result As String = Output
            Dim regex As New Regex(CALCULATION_REGEX, RegexOptions.Multiline)
            Dim index As Int32

            For Each pair As KeyValuePair(Of String, Object) In _valueList
                result = result.Replace(VARIABLE_OPEN & pair.Key & VARIABLE_CLOSE, pair.Value.ToString)
            Next

            index = result.IndexOf(CALCULATION_OPEN)

            Do While index >= 0
                Dim length As Int32 = 2
                Dim expression As String = String.Empty
                If index >= 0 Then
                    Dim parenthesis As Int32 = 1
                    Do While parenthesis > 0
                        Select Case result.Substring(index + length, 1)
                            Case VARIABLE_OPEN
                                parenthesis += 1
                            Case VARIABLE_CLOSE
                                parenthesis -= 1
                        End Select
                        If Not parenthesis = 0 Then length += 1
                    Loop
                    expression = result.Substring(index + 2, length - 2)
                End If
                result = result.Stuff(index, length + 1, CStr(_calculator.Evaluate(expression)))
                index = result.IndexOf(CALCULATION_OPEN)
            Loop

            Return result
        End Function
#End Region

#Region "Event Handlers"
        Private Sub Tables_EvaluateParameter(ByVal sender As Object, ByVal args As EvaluateTableParameterArgs)
            Dim key As String = args.Name

            If Not key.Contains(".") Then
                Dim tableName As String = DirectCast(sender, TableGrammarTable).Name
                key = tableName & "." & args.Name
            End If

            If _valueList.ContainsKey(key) Then
                args.Args.Result = _valueList(key)
            ElseIf ParameterExists(args.Name) Then
                args.Args.Result = Parameter(args.Name)
            End If
        End Sub

        Private Sub Tables_EvaluteFunction(sender As Object, e As EvaluateTableFunctionArgs)

        End Sub

        Private Sub _calculator_EvaluateParameter(name As String, args As ParameterArgs) Handles _calculator.EvaluateParameter
            If _valueList.ContainsKey(name) Then
                args.Result = _valueList(name)
            End If
        End Sub
#End Region

    End Class
End Namespace

File: Grammars/Table/TableGrammar.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Xml.Serialization;
using NCalc;
using Utility;

namespace Grammars.Table
{
    [XmlRoot(Namespace = "", IsNullable = false)]
    public class TableGrammar : BaseGrammar
    {
        private const string CALCULATION_REGEX = "\\[=.*?\\]";
        private const char VARIABLE_OPEN = '[';
        private const char VARIABLE_CLOSE = ']';
        private const string CALCULATION_OPEN = "[=";

        private Calculator _calculator;


        [XmlArray("tables")]
        [XmlArrayItem("table", IsNullable = false)]
        private ObservableCollection<TableGrammarTable> _Tables =
            new ObservableCollection<TableGrammarTable>();


        private Dictionary<string, object> _valueList = new Dictionary<string, object>();

        public TableGrammar()
        {
            _calculator = new Calculator(this);
            _calculator.EvaluateParameter += _calculator_EvaluateParameter;
        }

        private ObservableCollection<TableGrammarTable> Tables
        {
            get { return _Tables; }
            set { _Tables = value; }
        }

        [XmlElement("output")]
        public string Output { get; set; }


        public override string Analyze()
        {
            return string.Empty;
        }


        protected override string GenerateName()
        {
            _valueList = new Dictionary<string, object>();
            foreach (var table in Tables)
            {
                table.EvaluateParameter += Tables_EvaluateParameter;
                table.EvaluateFunction += Tables_EvaluteFunction;
                var newResults = table.ProcessTable();
                foreach (var result in newResults)
                {
                    _valueList.Add(table.Name + "." + result.Key, result.Value);
                }
                table.EvaluateParameter -= Tables_EvaluateParameter;
                table.EvaluateFunction -= Tables_EvaluteFunction;
            }
            return FillOutput();
        }


        private string FillOutput()
        {
            var result = Output;

            foreach (var pair in _valueList)
            {
                result = result.Replace(
                    VARIABLE_OPEN + pair.Key + Convert.ToString(VARIABLE_CLOSE),
                    pair.Value.ToString());
            }

            var index = result.IndexOf(CALCULATION_OPEN);

            while (index >= 0)
            {
                var length = 2;
                var expression = string.Empty;
                if (index >= 0)
                {
                    var parenthesis = 1;
                    while (parenthesis > 0)
                    {
                        var res = result.Substring(index + length, 1);

                        if (res == VARIABLE_OPEN.ToString())
                        {
                            parenthesis++;
                        }
                        else if (res == VARIABLE_CLOSE.ToString())
                        {
                            parenthesis--;
                        }
                        if (parenthesis != 0)
                        {
                            length++;
                        }
                    }
                    expression = result.Substring(index + 2, length - 2);
                }
                result = result.Stuff(index, length + 1, (string) (_calculator.Evaluate(expression)).ToString());
                index = result.IndexOf(CALCULATION_OPEN);
            }

            return result;
        }


        private void Tables_EvaluateParameter(object sender, TableGrammarTable.EvaluateTableParameterArgs args)
        {
            var key = args.Name;

            if (!key.Contains("."))
            {
                var tableName = ((TableGrammarTable) sender).Name;
                key = tableName + "." + args.Name;
            }

            if (_valueList.ContainsKey(key))
            {
                args.Args.Result = _valueList[key];
            }
            else if (ParameterExists(args.Name))
            {
                args.Args.Result = get_Parameter(args.Name);
            }
        }

        private void Tables_EvaluteFunction(object sender, TableGrammarTable.EvaluateTableFunctionArgs e)
        {
        }

        private void _calculator_EvaluateParameter(string name, ParameterArgs args)
        {
            if (_valueList.ContainsKey(name))
            {
                args.Result = _valueList[name];
            }
        }
    }
}